src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIAssistantClient.cs

```diff
@@ -0,0 +1,350 @@
+        {
+            updates = _assistantClient.SubmitToolOutputsToRunStreamingAsync(threadId, existingRunId, toolOutputs, cancellationToken);
+        }
+        else if (threadId is null)
+        {
+            ThreadCreationOptions creationOptions = new();
+            foreach (var message in runOptions.AdditionalMessages)
+            {
+                creationOptions.InitialMessages.Add(message);
+            }
+
+            runOptions.AdditionalMessages.Clear();
+
+            updates = _assistantClient.CreateThreadAndRunStreamingAsync(_assistantId, creationOptions, runOptions, cancellationToken: cancellationToken);
+        }
+        else
+        {
+            updates = _assistantClient.CreateRunStreamingAsync(threadId, _assistantId, runOptions, cancellationToken);
+        }
+
+        // Process each update.
+        await foreach (var update in updates.ConfigureAwait(false))
+        {
+            switch (update)
+            {
+                case MessageContentUpdate mcu:
+                    yield return new()
+                    {
+                        ChatThreadId = threadId,
+                        RawRepresentation = mcu,
+                        Role = mcu.Role == MessageRole.User ? ChatRole.User : ChatRole.Assistant,
+                        Text = mcu.Text,
+                    };
+                    break;
+
+                case ThreadUpdate tu when options is not null:
+                    threadId ??= tu.Value.Id;
+                    break;
+
+                case RunUpdate ru:
+                    threadId ??= ru.Value.ThreadId;
+
+                    StreamingChatCompletionUpdate ruUpdate = new()
+                    {
+                        AuthorName = ru.Value.AssistantId,
+                        ChatThreadId = threadId,
+                        CompletionId = ru.Value.Id,
+                        CreatedAt = ru.Value.CreatedAt,
+                        ModelId = ru.Value.Model,
+                        RawRepresentation = ru,
+                        Role = ChatRole.Assistant,
+                    };
+
+                    if (ru.Value.Usage is { } usage)
+                    {
+                        ruUpdate.Contents.Add(new UsageContent(new()
+                        {
+                            InputTokenCount = usage.InputTokenCount,
+                            OutputTokenCount = usage.OutputTokenCount,
+                            TotalTokenCount = usage.TotalTokenCount,
+                        }));
+                    }
+
+                    if (ru is RequiredActionUpdate rau && rau.ToolCallId is string toolCallId && rau.FunctionName is string functionName)
+                    {
+                        ruUpdate.Contents.Add(
+                            new FunctionCallContent(
+                                JsonSerializer.Serialize(new[] { ru.Value.Id, toolCallId }, OpenAIJsonContext.Default.StringArray!),
```

Comment:
I got rid of that additional Context property and am following the suggested approach of putting the run ID into the tool call ID.